File: android/app/src/main/java/com/mdreader/render/SvgGuard.kt
package com.mdreader.render

/**
 * Guards inline `<svg>…</svg>` blocks from marked's HTML-block parsing.
 *
 * marked follows CommonMark's HTML-block rule that ends an HTML block at the
 * first blank line. Inline SVGs almost always contain blank lines (between
 * `<defs>`, `<g>`, `<text>` groups), so without guarding marked terminates
 * the block mid-SVG and re-parses the remainder as markdown, corrupting the
 * diagram — on real documents a ~15 KB SVG collapsed to ~2.3 KB in the DOM
 * and rendered at 0×0.
 *
 * [guard] lifts every top-level SVG out of the markdown and replaces it with
 * a compact, blank-line-free placeholder. The renderer re-injects the
 * originals after `marked.parse` via `SourceBridge.getSvg` (see render.js),
 * so each SVG enters the DOM intact and verbatim.
 *
 * SVGs nested inside fenced code blocks are left in place: marked renders
 * those verbatim, and guarding them would corrupt the code display.
 *
 * Pure and side-effect free: unit-testable on the JVM without a WebView.
 * The fence handling mirrors [MermaidFenceNormalizer] so fenced code is
 * respected by both passes.
 */
object SvgGuard {
    private val fenceRegex =
        Regex("""^([ \t]{0,3})(`{3,}|~{3,})[ \t]*([\w-]+)?[ \t]*(\{.*\})?[ \t]*$""")
    private val svgRegex = Regex("""<svg\b[\s\S]*?</svg>""")

    const val MARKER = '\u0001'
    const val END = '\u0002'

    data class Guarded(val markdown: String, val svgs: List<String>)

    /** Placeholder injected for the [index]-th SVG; matched by render.js. */
    fun placeholder(index: Int): String = "$MARKER${index}$END"

    /**
     * Returns [markdown] with every top-level (non-fenced) `<svg>…</svg>` run
     * replaced by a placeholder, together with the extracted SVG texts in
     * insertion order. Returns the input unchanged (with an empty list) when
     * it contains no `<svg` token at all.
     */
    fun guard(markdown: String): Guarded {
        if (!markdown.contains("<svg")) return Guarded(markdown, emptyList())
        val svgs = mutableListOf<String>()
        val lines = markdown.split('\n')
        val out = StringBuilder()
        var i = 0
        var inFence = false
        var fenceMarker = ""
        while (i < lines.size) {
            val line = lines[i]
            val fm = fenceRegex.matchEntire(line.trimEnd())
            if (fm != null) {
                val marker = fm.groupValues[2]
                if (!inFence) {
                    inFence = true
                    fenceMarker = marker
                } else if (marker.isNotEmpty() && fenceMarker.isNotEmpty() &&
                    marker[0] == fenceMarker[0] && marker.length >= fenceMarker.length
                ) {
                    inFence = false
                    fenceMarker = ""
                }
                out.append(line).append('\n')
                i++
                continue
            }
            if (inFence) {
                out.append(line).append('\n')
                i++
                continue
            }
            // Top-level: collect an SVG block from the first line containing
            // `<svg` up to the next line that closes it, then extract every
            // <svg>…</svg> run inside (usually one).
            if ("<svg" in line) {
                val buf = StringBuilder(line)
                var j = i
                if ("</svg>" !in line) {
                    j = i + 1
                    while (j < lines.size) {
                        buf.append('\n').append(lines[j])
                        if ("</svg>" in lines[j]) break
                        j++
                    }
                }
                val replaced = svgRegex.replace(buf.toString()) {
                    svgs.add(it.value)
                    placeholder(svgs.size - 1)
                }
                out.append(replaced).append('\n')
                i = j + 1
                continue
            }
            out.append(line).append('\n')
            i++
        }
        return Guarded(out.toString().removeSuffix("\n"), svgs)
    }
}

File: macos/MDreader/render/SvgGuard.swift
import Foundation

struct Guarded {
    let markdown: String
    let svgs: [String]
}

enum SvgGuard {
    static let marker: Character = "\u{01}"
    static let end: Character = "\u{02}"

    private static let svgRegex: NSRegularExpression = {
        try! NSRegularExpression(pattern: "<svg\\b[\\s\\S]*?</svg>", options: [])
    }()

    static func placeholder(_ index: Int) -> String {
        "\(marker)\(index)\(end)"
    }

    static func protect(_ markdown: String) -> Guarded {
        if !markdown.contains("<svg") {
            return Guarded(markdown: markdown, svgs: [])
        }
        var svgs: [String] = []
        let lines = markdown.components(separatedBy: "\n")
        var out = ""
        var i = 0
        var inFence = false
        var fenceMarker = ""
        while i < lines.count {
            let line = lines[i]
            if let fm = Fence.match(trimTrailingWhitespace(line)) {
                let markerRun = fm.marker
                if !inFence {
                    inFence = true
                    fenceMarker = markerRun
                } else if !markerRun.isEmpty && !fenceMarker.isEmpty &&
                    markerRun.first == fenceMarker.first &&
                    markerRun.count >= fenceMarker.count {
                    inFence = false
                    fenceMarker = ""
                }
                out += line + "\n"
                i += 1
                continue
            }
            if inFence {
                out += line + "\n"
                i += 1
                continue
            }
            if line.contains("<svg") {
                var buf = line
                var j = i
                if !line.contains("</svg>") {
                    j = i + 1
                    while j < lines.count {
                        buf += "\n" + lines[j]
                        if lines[j].contains("</svg>") { break }
                        j += 1
                    }
                }
                let replaced = extractSvgs(buf, into: &svgs)
                out += replaced + "\n"
                i = j + 1
                continue
            }
            out += line + "\n"
            i += 1
        }
        if out.hasSuffix("\n") { out.removeLast() }
        return Guarded(markdown: out, svgs: svgs)
    }

    private static func extractSvgs(_ text: String, into svgs: inout [String]) -> String {
        let ns = text as NSString
        let range = NSRange(location: 0, length: ns.length)
        let matches = svgRegex.matches(in: text, range: range)
        var result = ""
        var cursor = 0
        for m in matches {
            let mr = m.range
            result += ns.substring(with: NSRange(location: cursor, length: mr.location - cursor))
            svgs.append(ns.substring(with: mr))
            result += placeholder(svgs.count - 1)
            cursor = mr.location + mr.length
        }
        result += ns.substring(with: NSRange(location: cursor, length: ns.length - cursor))
        return result
    }
}
